Split change outputs with asset quantities exceeding the maximum

jonathanknowles opened the pull request:
# Issue Number

https://github.com/input-output-hk/cardano-wallet/issues/2532
ADP-726

# Overview

Although quantities of individual assets are effectively unlimited, a transaction on the blockchain can never include an asset quantity greater than `maxBound :: Word64`.

This PR tweaks the coin selection algorithm to detect change bundles containing excessively large asset quantities.

If such a change bundle is detected, we now split the change bundle up into smaller bundles using [equipartitioning](https://en.wiktionary.org/wiki/equipartition).

# Example

Let's suppose the maximum allowable token quantity is **`5`**, and we have a change map with the following quantities:

```haskell
[("a", 11), ("b", 5), ("c", 12)]
```

In this case, we must divide the map into **_at least three_** smaller maps in order to not exceed the maximum allowable token quantity.

Under the equipartitioning scheme, this would give us:

```haskell
[("a", 3), ("b", 1), ("c", 4)]
[("a", 4), ("b", 2), ("c", 4)]
[("a", 4), ("b", 2), ("c", 4)]
```

Note that while the overall sum is preserved, the individual bundles are almost equal, **_but not quite_**: this is because `11` and `5` are not divisible by `3`. We must therefore accept a small loss of proportionality in the result.

# Details

An **_equipartition_** of a bundle **_b_** is a _partition_ into multiple bundles, where for every asset **_a_** in the set of assets contained in **_b_**, the difference between the following quantities is either _zero_ or _one_ :

- The smallest quantity of asset **_a_** in the resultant bundles
- The greatest quantity of asset **_a_** in the resultant bundles

In order to determine the number of parts in which to split a given bundle, we choose the **_smallest_** number of parts that still allows us satisfy the goal of not exceeding the maximum allowable quantity in any given bundle.

# Performance

In order to avoid evaluating a partition for every single change output, we **_short circuit_** in the event that there is no token quantity greater than the maximum allowable quantity:

```haskell
equipartitionTokenMapWithMaxQuantity m (TokenQuantity maxQuantity)
    | maxQuantity == 0 =
        maxQuantityZeroError
    | currentMaxQuantity <= maxQuantity =
        m :| []
    | otherwise =
        equipartitionTokenMap m (() :| replicate extraPartCount ())
```

# Testing

## Property tests

Equipartitioning behaviour is tested by the following property tests:
- `prop_equipartitionTokenBundle*`
- `prop_equipartitionTokenMap*`

## Unit tests

As a sanity check, this PR also provides unit tests for `performSelection` with inputs containing token quantities that are close to the maximum, and demonstrates that change bundles are correctly partitioned in the results.

jonathanknowles committed:
Add function `equipartitionTokenMap`.

jonathanknowles committed:
Add function `unsafePartitionCoin`.

jonathanknowles committed:
Add function `unsafePartitionNatural`.

Use this function to replace the common functionality of:

- `unsafePartitionCoin`
- `unsafePartitionTokenQuantity`

jonathanknowles committed:
Add function `equipartitionTokenMapWithMaxQuantity`.

jonathanknowles committed:
Add function `equipartitionCoin`.

jonathanknowles committed:
Add function `equipartitionTokenBundle`.

jonathanknowles committed:
Add function `equipartitionTokenBundleWithMaxQuantity`.

jonathanknowles committed:
When making change, split up any change maps with excessive token quantities.

jonathanknowles committed:
Consolidate documentation for equipartitioning functions.

These functions share the same general properties, so it makes sense to
document these properties in a single place.

jonathanknowles committed:
Add unit tests for the maximum token quantity.

This change adds unit tests for `performSelection` to demonstrate what
should happen when inputs are selected whose total quantity of a given
asset is close to the maximum token quantity.

It should not be possible to produce a `SelectionResult` where one or
more change outputs contains a token quantity greater than the maximum
(maxBound :: Word64).

These unit tests serve as a simple sanity check: the primary testing of
the change map partitioning behaviour is handled by the following
property tests:

 - prop_equipartitionCoin_*
 - prop_equipartitionTokenBundle_*
 - prop_equipartitionTokenMap_*

jonathanknowles committed:
Extract out function `equipartitionTokenQuantity`.

This function is useful for writing unit tests.

jonathanknowles committed:
Use `equipartitionNatural` to simplify equipartitioning functions.

rvl commented on .hlint.yaml:
```diff
@@ -54,6 +54,7 @@
 - ignore: {name: "Redundant multi-way if"} # Using this style might be appropriate given the circumstances.
 - ignore: {name: "Use const"} # Sometimes using a lambda expression is simpler and clearer.
 - ignore: {name: "Redundant id"} # Sometimes this is useful for vertical alignment of elements.
+- ignore: {name: "Use camelCase"} # Sometimes this is useful in test code.
```
I think we normally want this hlint enabled for application code.

jonathanknowles replied:
Fair enough!

I've moved the suppression of this warning to just the `RoundRobinSpec` module.

Fixed in: b9454787a8184e0e83338602147614f56763c5bd

rvl commented on lib/core/src/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobin.hs:
```diff
@@ -1219,6 +1436,16 @@ newtype AssetCount a = AssetCount
+--   a transaction.
+--
+maxTokenQuantity :: TokenQuantity
+maxTokenQuantity = TokenQuantity $ fromIntegral (maxBound :: Word64)
```
It's a bit of a mouthful, but renaming maxTokenQuantity -> maxTokenQuantityInTxOut will prevent misleading any readers.

jonathanknowles replied:
I agree this is a good idea. 👍🏻 

Fixed in 0cfb86f7ffc555d740944bbac506e2676a73f701

rvl commented on lib/core/src/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobin.hs:
```diff
@@ -855,11 +870,25 @@ makeChange minCoinFor requiredCost mExtraCoinSource inputBundles outputBundles
+        -- split these maps up along with their corresponding output coins:
+        & splitMapsWithExcessiveQuantities
+      where
+        splitMapsWithExcessiveQuantities
+            :: NonEmpty (TokenMap, Coin) -> NonEmpty (TokenMap, Coin)
+        splitMapsWithExcessiveQuantities =
```
I would prefer if the unit testing were based on properties of this function.

jonathanknowles replied:
> I would prefer if the unit testing were based on properties of this function.

Fair enough.

Since the critical part of this function relates to list-processing, I've extracted this out as a separate function, for which I've added some simple property tests.

Now the `splitMapsWithExcessiveQuantities` function is simply a wrapper function that bundles and unbundles the `(TokenMap, Coin)` pair.

See 9cf72cba44204032a1238c40e4c7ee156ad6914d.

rvl commented on lib/core/src/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobin.hs:
```diff
@@ -1145,6 +1174,194 @@ makeChangeForCoin targets excess =
+    -- ^ Represents the number of portions in which to partition the coin.
+    -> NonEmpty Coin
+    -- ^ The partitioned coins.
+equipartitionCoin c =
```
Given the max supply of ada, there's no need to apply this fix to Coin.

jonathanknowles replied:
> Given the max supply of ada, there's no need to apply this fix to Coin.

The point here is that we need to divide output coins **_together_** with change maps, so that we can retain proportionality. (Each change map is paired with an output coin, so that when we assign coins to change maps, we can assign them in proportion to the sizes of the coins in the original outputs.)

Let's say we need to partition the following token map into two:

- TokenMap A:
    - Banana: 100
    - Apple: 10

This would give us two token maps with the following tokens:

- TokenMap A1:
    - Banana: 50
    - Apple: 5
- TokenMap A2:
    - Banana: 50
    - Apple: 5

However, the change generation algorithm pairs each change token map with a coin from the original outputs, so that when we eventually assign ada values to change maps (to make change bundles), we can assign them in proportion to the sizes of the ada in the original outputs.

So we also need to divide the coin value.

Let's say this particular change map was paired with an output coin value of `100`. That will mean that each divided change map will be paired with a coin value of `50`.

For the sake of convenience, when dividing the pair (change map, output coin value), we treat it as a token bundle, so that we can use the `equipartitionTokenBundleWithMaxQuantity` function.

rvl commented on lib/core/src/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobin.hs:
```diff
@@ -1145,6 +1174,194 @@ makeChangeForCoin targets excess =
+    -- ^ Represents the number of portions in which to partition the bundle.
+    -> NonEmpty TokenBundle
+    -- ^ The partitioned bundle.
+equipartitionTokenBundle (TokenBundle c m) count =
```
This function shouldn't be needed, right? Only equipartitionTokenMap.

jonathanknowles replied:
> This function shouldn't be needed, right? Only equipartitionTokenMap.

You're right -- I've deleted it.

Fixed in 607ec07474ba813e9e7adc344c58c28b7f8f8fac

rvl reviewed (commented):
This will work but it was confusing to review and seems like a lot of code for something which is basically this:
```haskell
example :: [(String, Natural)]
example = [("a", 11), ("b", 5), ("c", 12)]

-- >>> limitSize 5 example
-- [[("a",5),("b",5),("c",5)],[("a",5),("c",5)],[("a",1),("c",2)]]
limitSize :: (Traversable t, Integral n) => n -> [t n] -> [[t n]]
limitSize lim = group . map (fmap (splitNum lim))
  where
    group = L.transpose . map sequence

splitNum :: Integral n => n -> n -> [n]
splitNum q a = replicate (fromIntegral n) q ++ if r == 0 then [] else [r]
  where
    n = a `div` q
    r = a `mod` q
```
So is it possible to adjust this limitSize slightly so that it can be applied to TokenMap?

jonathanknowles committed:
Remove function `equipartitionTokenMap`.

This function is now redundant.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/2536#discussion_r583377781

jonathanknowles committed:
Rename `maxTokenQuantity` to `maxTxOutTokenQuantity`.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/2536#discussion_r583376580

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobinSpec.hs:
```diff
@@ -1576,6 +1838,194 @@ unit_makeChangeForUserSpecifiedAsset =
+prop_equipartitionNatural_fair n count =
+    prop_equipartitionNatural_fair_inner $ equipartitionNatural n count
+
+prop_equipartitionNatural_fair_inner :: NonEmpty Natural -> Property
```
This function can actually now be inlined into `prop_equipartitionNatural_fair`.

jonathanknowles committed:
Inline function `prop_equipartitionNatural_fair_inner`.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/2536/files#r583385917

jonathanknowles committed:
Suppress hlint "Use camelCase" warning for `RoundRobinSpec`.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/2536#discussion_r583376024

jonathanknowles committed:
Extract out list-processing part of `splitMapsWithExcessiveQuantities`.

This change extracts out the list-processing part of
`splitMapsWithExcessiveQuantities` into a separate function
`equipartitionTokenBundlesWithMaxQuantity` that can be tested.

It also adds property tests for this function.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/2536#discussion_r583376889

KtorZ commented on lib/core/src/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobin.hs:
```diff
@@ -1145,6 +1176,195 @@ makeChangeForCoin targets excess =
+    TokenQuantity currentMaxQuantity = TokenMap.maximumQuantity m
+
+    extraPartCount :: Int
+    extraPartCount = floor $ pred currentMaxQuantity % maxQuantity
```
That's a slightly odd way to compute this and quite confusing at first. But ok with the construction of the non-empty list above :heavy_check_mark: 

That'd be clearer with a let-binding because this calculation only makes sense with the surrounding context :smirk:

jonathanknowles replied:
> That's a slightly odd way to compute this and quite confusing at first. But ok with the construction of the non-empty list above heavy_check_mark

Yes, it does seem odd at first. Though you're absolutely right -- it's this way because of the non-empty list.

The test case for this uses a more straightforward calculation:

```hs
-- | Computes the number of parts that 'equipartitionTokenMapWithMaxQuantity'
--   should return.
--
equipartitionTokenMapWithMaxQuantity_expectedLength
    :: TokenMap -> TokenQuantity -> Int
equipartitionTokenMapWithMaxQuantity_expectedLength
    m (TokenQuantity maxQuantity) =
        max 1 $ ceiling $ currentMaxQuantity % maxQuantity
  where
    TokenQuantity currentMaxQuantity = TokenMap.maximumQuantity m
```

KtorZ commented on lib/core/test/unit/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobinSpec.hs:
```diff
@@ -1008,6 +1084,203 @@ prop_coinSelectionLens_givesPriorityToCoins (Blind (Small u)) =
+      , (Coin 1_000_000, [(assetA, q2)])
+      ]
+    boundaryTestChange =
+      [ (Coin 500_000, [(assetA, maxTxOutTokenQuantity)]) ]
```
:heavy_check_mark:

KtorZ commented on lib/core/test/unit/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobinSpec.hs:
```diff
@@ -1008,6 +1084,203 @@ prop_coinSelectionLens_givesPriorityToCoins (Blind (Small u)) =
+      , (Coin 1_000_000, [(assetA, q2)])
+      ]
+    boundaryTestChange =
+      [ (Coin 500_000, [(assetA, maxTxOutTokenQuantity)]) ]
```
:heavy_check_mark:

KtorZ commented on lib/core/test/unit/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobinSpec.hs:
```diff
@@ -1008,6 +1084,203 @@ prop_coinSelectionLens_givesPriorityToCoins (Blind (Small u)) =
+    boundaryTestChange =
+      [ (Coin 250_000, [(assetA, q1)])
+      , (Coin 250_000, [(assetA, q2)])
+      ]
```
:heavy_check_mark:

KtorZ commented on lib/core/test/unit/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobinSpec.hs:
```diff
@@ -1008,6 +1084,203 @@ prop_coinSelectionLens_givesPriorityToCoins (Blind (Small u)) =
+    boundaryTestChange =
+      [ (Coin 250_000, [(assetA, maxTxOutTokenQuantity)])
+      , (Coin 250_000, [(assetA, maxTxOutTokenQuantity)])
+      ]
```
:heavy_check_mark:

KtorZ commented on lib/core/test/unit/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobinSpec.hs:
```diff
@@ -1008,6 +1084,203 @@ prop_coinSelectionLens_givesPriorityToCoins (Blind (Small u)) =
+      [ (Coin 250_000, [(assetA, maxTxOutTokenQuantity)])
+      , (Coin 250_000, [(assetA, maxTxOutTokenQuantity)])
+      ]
+
```
It'd have been nice to also show an example where some assets are specified in the outputs, but the resulting change is still going above limit.

KtorZ reviewed (approved):
bors merge
